engine/src/flutter/display_list/geometry/dl_path.h

```diff
@@ -15,6 +17,38 @@ namespace flutter {
 using DlPathFillType = impeller::FillType;
 using DlPathBuilder = impeller::PathBuilder;
 
+/// @brief   Collection of functions to receive path segments from the
+///          underlying path representation via the DlPath::Dispatch method.
+///
+/// The conic_to function is optional. If the receiver understands rational
+/// quadratic Bezier curve forms then it should accept the curve parameters
+/// and return true, otherwise it can return false and the dispatcher will
+/// provide the path segment in a different form via the other methods.
+///
+/// The dispatcher might not call the recommend_size or recommend_bounds
+/// functions if the original path does not contain such information.
+///
+/// The dispatcher will always call the path_info function, though the
+/// is_convex parameter may be conservatively reported as false if the
+/// original path does not contain such info.
+struct DlPathReceiver {
+  std::function<void(size_t verb_count, size_t point_count)> recommend_size =
+      [](size_t, size_t) {};
+  std::function<void(const DlRect& bounds)> recommend_bounds =
+      [](const DlRect& bounds) {};
+  std::function<void(DlPathFillType fill_type, bool is_convex)> path_info;
```

Comment:
If there was a way to write this so that they must provide an implementation for all of these fields that don't have a default value, I'd like to use that. Unfortunately, std::function has a default initializer defined that basically crashes when you call it (how completely unuseful is that?) so that if an implementor doesn't provide one of the required methods, they have a little bit of a puzzle on their hands to debug.

Reply:
Note there is no warning in VSCode or during compilation that the field was uninitialized.

Reply:
Why do this with callbacks versus accepting a visitor interface?

Reply:
This can be implemented within a function using lambdas. Most Visitor patterns I've seen involve having to define classes and templates which would need to be done outside the function which spreads out the implementation, but maybe I'm not thinking of the same paradigm that you are suggesting?

Reply:
I would defer to @gaaclarke on this as he is the visitor pattern expert, but I think in this case you could more or less write a struct inline that implements the interface.

Reply:
Why aren't you just using virtual methods?

```c++
class DlPathReceiver {
  public:
    virtual void MoveTo(const DlPoint& p2) = 0;
};

class AndroidPathReceiver : public DlPathReceiver {
  public:
    void MoveTo(const DlPoint& p2) override {
      env_->CallVoidMethod(androidPath_, path_move_to_method,  //
                                        p2.x, p2.y);
    }
};

class CGReceiver : public DlPathReciever {
   void MoveTo(const DlPoint& p2) override {
       CGPathMoveToPoint(pathRef_, nil, p2.x, p2.y);
   },
}; 

```

Reply:
That would require defining a class somewhere to implement them. The functions allow in-line lambda definitions to keep the source footprint smaller. I could go either way on that.

Reply:
See https://github.com/flutter/flutter/pull/164753/files#diff-0b4c010fe33223d6d245c11d1327f10cb33b48b7d22ba709f429c978b6561a68

Reply:
Written either way DlPathReceiver takes up about the same amount of space.  You can just throw the implementations in an anonymous namespace in the cc file you need them in.  I don't think it will be more code if your super class defaults to doing nothing (I think that's what you are saying you want).

Reply:
And virtual methods provide that assurance that someone will not fail to implement the required methods. OK, I'll make that change...

Reply:
Switched to a virtual class/methods...